File: src/python/graph/2_remove_degree2_nets.py
# file: 2_remove_degree2_nets.py
# Remove degree-2 NET nodes and connect neighbors directly

import networkx as nx # read_gml, write_gml
import sys            # argv

def main():
    input_file, output_file = sys.argv[1], sys.argv[2]
    G = nx.read_gml(input_file)
    H = G.copy()
    processed = set()
    while True:
        found = False
        for node in list(H.nodes()):
            if node in processed: continue
            if H.nodes[node].get('type') == 'net' and H.degree(node) == 2:
                n1, n2 = list(H.neighbors(node))
                if not H.has_edge(n1, n2): H.add_edge(n1, n2)
                H.remove_node(node)
                processed.add(node)
                found = True
                break
            processed.add(node)
        if not found: break
    nx.write_gml(H, output_file)

if __name__ == "__main__":
    main()

File: src/typescript/graph/2_remove_degree2_nets.ts
import fs from 'fs';
import Graph from 'graphology';

export function removeDegree2Nets(inputPath: string, outputPath: string): Graph {
    console.log(`[GRAPH] 2. Removing degree-2 nets from ${path.basename(inputPath)}`);
    const G = new Graph();
    if (!fs.existsSync(inputPath)) return G;
    G.import(JSON.parse(fs.readFileSync(inputPath, 'utf8')));
    
    const H = G.copy();
    let removedCount = 0;
    while (true) {
        let found = false;
        H.forEachNode((node, attr) => {
            if (found) return;
            if (attr.type === 'net' && H.degree(node) === 2) {
                const neighbors = H.neighbors(node);
                const n1 = neighbors[0];
                const n2 = neighbors[1];
                if (!H.hasEdge(n1, n2)) H.addEdge(n1, n2);
                H.dropNode(node);
                found = true;
                removedCount++;
            }
        });
        if (!found) break;
    }
    console.log(`[GRAPH] Removed ${removedCount} degree-2 nets.`);
    fs.writeFileSync(outputPath, JSON.stringify(H.export()));
    return H;
}
